web/cobrands/fixmystreet/assets.js

```diff
@@ -16,8 +16,58 @@ var fixmystreet = fixmystreet || {};
     };
 })();
 
+OpenLayers.Layer.VectorAsset = OpenLayers.Class(OpenLayers.Layer.Vector, {
+    initialize: function(name, options) {
+        OpenLayers.Layer.Vector.prototype.initialize.apply(this, arguments);
+        // Might only be able to fill in fields once they've been returned from the server
+        $(fixmystreet).on('report_new:category_change:extras_received', this.update_layer_visibility.bind(this));
+        // But also want to do it immediately in case it's hiding the form or something
+        $(fixmystreet).on('report_new:category_change', this.update_layer_visibility.bind(this));
+    },
+
+    update_layer_visibility: function() {
+        if (!fixmystreet.map) {
+          return;
+        }
+
+        if (!this.fixmystreet.always_visible) {
+            // Show/hide the asset layer when the category is chosen
+            var category = $('#problem_form select#form_category').val();
+            if (fixmystreet.assets.check_layer_relevant(this.fixmystreet, category)) {
+                this.setVisibility(true);
+                if (this.fixmystreet.fault_layer) {
+                    this.fixmystreet.fault_layer.setVisibility(true);
+                }
+                this.zoom_to_assets();
+            } else {
+                this.setVisibility(false);
+                if (this.fixmystreet.fault_layer) {
+                    this.fixmystreet.fault_layer.setVisibility(false);
+                }
+            }
+        } else {
+            if (fixmystreet.bodies && this.fixmystreet.body) {
```

Comment:
My worry here is if the centre of the map is in one council, but part of a council which there are assets is being shown and presumably any road-based layers should be shown there. Otherwise you could have a slight case (perhaps a race condition) of clicking on a Bromley part of the map that is centred on Bexley, and it not finding the TfL road asset (though I guess when the map panned and it all loaded in, it might then kick off another 'find' of the asset then? [1]) But I do see we don't want to bother HTTP-requsting checking a Bromley layer for a map in Scotland...

[1] Hmm, I guess if the parent/grandparent change I suggest is made, then the visibility event will fire before the check the features event, so it might be okay, though what if the assets need to be loaded from the server? Would be good to confirm this more robustly somehow!

Reply:
Given we don't show most assets until they click the pin and select a category I think this problem is largely moot. Yes, we might then need to load a road layer but with the parent/grandparent change that problem is fixed I think.